secp256k1-sys/README.md

```diff
@@ -24,12 +28,12 @@ $ ./vendor-libsecp.sh depend <version-code> <rev>
 ```
 
 - Where `<version-code>` is the secp256k1-sys version number underscored: `0_1_2`.
-- Where `<rev>` is the git revision of libsecp256k1 to checkout.
+- Where `<rev>` is the git revision of `libsecp256k1` to checkout.
 
 
 ## Linking to external symbols
 
-If you want to compile this library without using the bundled symbols (which may
-be required for integration into other build systems), you can do so by adding
-`--cfg=rust_secp_no_symbol_renaming'` to your `RUSTFLAGS` variable.
+If you want to compile this library without using the bundled symbols (which may be required for
+integration into other build systems), you can do so by adding `--cfg=rust_secp_no_symbol_renaming'`
+to your `RUSTFLAGS` variable.
 
```

Comment:
Maybe we could add the information about the limitations of this @apoelstra mentioned while we're at it. I think this should be correct. I suggest putting it right after the heading.

> **Danger: doing this incorrectly may have catastrophic consequences!**
> This is mainly intended for applications consisting of various programming languages that intend to link the same library to save space or **bundles** of multiple binaries coming from the same source.
> Do **not** use this to link to a random `secp256k1` library you found in your OS!
> Packaging the library into a separate distribution package could be done with care.
> E.g. in Debian if you'd have `libsecp256k1` with version `1.2.3` then all the binaries using it must use `>= 1.2.3, << 1.2.4`, assuming you will only ever release binary-compatible versions of the package.
> This is because there are no binary compatibility guarantees and the library is security-critical.
> Note also that unless you're packaging the library for an official repository you should prefix your package and the library with a string specific to you. E.g. if you have a set of packages called `my-awesome-packages` you should package `libsecp256k1` as `libmy-awesome-packages-secp256k1` and depend on that library/package name from your application.
> If you do any of this you must make sure that either every component that depends on `libsecp256k` is written to depend on the same version (`-sys` crate **not** being duplicated in case of Rust) or that the differing components use static bundled version with the different prefixes.

Reply:
>E.g. in Debian if you'd have libsecp256k1 with version 1.2.3 then all the binaries using it must use >= 1.2.3, << 1.2.4, assuming you will only ever release binary-compatible versions of the package. This is because there are no binary compatibility guarantees and the library is security-critical.

I don't understand this sentence -- there is only one release of libsecp in existence and it is not in any distribution AFAIK. What distributions have actually packaged is random shit. I'm also not sure why, "if there are no binary compatibility guarantees", merely constraining a version number would provide any value.

I'm also not totally sure who "you" is here -- is in the author of the Rust code, or a distribution maintainer, or a libsecp developer, or what?

Other than this, this text looks good.

Reply:
> and it is not in any distribution AFAIK

Indeed, this is written in case someone wants to package it. But now I'm thinking it may overlooked anyway. I guess some doc like that should go to the upstream library instead.

> I'm also not sure why, "if there are no binary compatibility guarantees", merely constraining a version number would provide any value.

If you control *all* the packages and make sure you won't release version 1.2.3-2 that is binary-incompatible with 1.2.3-1 then it's OK to update the library. (e.g. if you only made cosmetic changes like documentation or fixed something minor inside a function body)

Basically, if you control the library (package) and make sure it doesn't get confused/swapped with something else then it's OK to link to it from a binary you control.

> I'm also not totally sure who "you" is here

The author of the binary using `libsecp256k1` and potentially distribution maintainer. (I think they are sometimes the same person? I do make my own packages for some of my own projects but that doesn't make me distribution maintainer...)

Reply:
How about `If you are packaging software that depends on rust-secp256k1, using this flag to link to another package, make sure you stay within the binary compatibility guarantees of that package. For example, in Debian if you need libsecp256k1 1.2.4, make sure your package requires a version strictly >= 1.2.3 << 1,2.4.`

Reply:
Nice!